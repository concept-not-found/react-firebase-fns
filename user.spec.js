const {createElement} = require('react')
const {create} = require('react-test-renderer')

const FirebaseMock = require('./firebase-mock')
const UserFactory = require('./user')

describe('user', () => {
  it('should render undefined initially', () => {
    const {Firebase} = FirebaseMock()
    const User = UserFactory(Firebase)
    const component = create(createElement(User.Provider, {},
      createElement(User.Consumer, {},
        (user) => `user: ${JSON.stringify(user)}`
      )
    ))
    expect(component.toJSON())
      .toMatchSnapshot()
  })

  it('when the user logs in, component is re-rendered', () => {
    const {onAuthStateChangedMock, Firebase} = FirebaseMock()
    const User = UserFactory(Firebase)
    const component = create(createElement(User.Provider, {},
      createElement(User.Consumer, {},
        (user) => `user: ${JSON.stringify(user)}`
      )
    ))
    const listener = onAuthStateChangedMock.mock.calls[0][0]
    listener({
      displayName: 'some user'
    })
    expect(component.toJSON())
      .toMatchSnapshot()
  })

  it('when the user logs out, component is re-rendered', () => {
    const {onAuthStateChangedMock, Firebase} = FirebaseMock()
    const User = UserFactory(Firebase)
    const component = create(createElement(User.Provider, {},
      createElement(User.Consumer, {},
        (user) => `user: ${JSON.stringify(user)}`
      )
    ))
    const listener = onAuthStateChangedMock.mock.calls[0][0]
    listener({
      displayName: 'some user'
    })
    listener(undefined)
    expect(component.toJSON())
      .toMatchSnapshot()
  })

  it('when component is unmounted, listener is unsubscribed', () => {
    const {unsubscribeOnAuthStateChanged, Firebase} = FirebaseMock()
    const User = UserFactory(Firebase)
    const component = create(createElement(User.Provider, {},
      createElement(User.Consumer, {},
        (user) => `user: ${JSON.stringify(user)}`
      )
    ))
    component.unmount()
    expect(unsubscribeOnAuthStateChanged.mock.calls.length).toBe(1)
  })
})
