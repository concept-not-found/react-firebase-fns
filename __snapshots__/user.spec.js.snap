// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user should render undefined initially 1`] = `"user: undefined"`;

exports[`user when the user logs in, component is re-rendered 1`] = `"user: {\\"displayName\\":\\"some user\\"}"`;

exports[`user when the user logs out, component is re-rendered 1`] = `"user: undefined"`;
